### PR TITLE
Add Ripen and Finish Gen4 Berries

### PR DESCRIPTION
**Pull request by ghoulslash**
## Description
Adds:
- Custap Berry (and similarly, the quick claw activation message)
- Micle Berry
- Jaboca/Rowap Berries
- Kee/Maranga Berries
- Ripen for all berries
- Sitrus Berry heals 25% in battle
- Added immediate consumption for stat boosting and flavor-hp-restoring berries to ITEMEFFECTS_MOVE_END

Not handled:
- Flinging a berry (fling not yet added)

Starf:
ripen_starf

Custap Berry / Quick Claw:
custap quick_claw

Micle Berry:
micle

Jaboca Berry:
jaboca

Kee Berry:
kee

This also fixes berry activation after Recycle:
recycle

### CODE REVIEW

File: src/battle_util.c [hunk 1]
```diff
@@ -5104,10 +5104,10 @@ enum
 };
 
 // second argument is 1/X of current hp compared to max hp
-static bool32 HasEnoughHpToEatBerry(u32 battlerId, u32 hpFraction, u32 itemId)
+bool32 HasEnoughHpToEatBerry(u32 battlerId, u32 hpFraction, u32 itemId)
 {
     bool32 isBerry = (ItemId_GetPocket(itemId) == POCKET_BERRIES);
-
+
```
Comment (ultima-soul) [inline src/battle_util.c:5110]:
```suggestion

```

File: src/battle_util.c [hunk 2]
```diff
@@ -5191,10 +5222,71 @@ static u8 RandomStatRaiseBerry(u32 battlerId, u32 itemId)
...
+    return 0;
+}
+
+static u8 DamagedStatBoostBerryEffect(u8 battlerId, u8 statId, u8 split)
+{
+    if (IsBattlerAlive(battlerId)
+      && TARGET_TURN_DAMAGED
+      && gBattleMons[battlerId].statStages[statId] < MAX_STAT_STAGE
+      && !DoesSubstituteBlockMove(gBattlerAttacker, battlerId, gCurrentMove)
+      && GetBattleMoveSplit(gCurrentMove) == split)
```
Comment (ultima-soul) [inline src/battle_util.c:5274]:
```suggestion
     && TARGET_TURN_DAMAGED
     && gBattleMons[battlerId].statStages[statId] < MAX_STAT_STAGE
     && !DoesSubstituteBlockMove(gBattlerAttacker, battlerId, gCurrentMove)
     && GetBattleMoveSplit(gCurrentMove) == split)
```

File: src/battle_util.c [hunk 3]
```diff
@@ -5989,6 +6145,50 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
                     gBattleScripting.statChanger = SET_STATCHANGER(STAT_SPATK, 1, FALSE);
                 }
                 break;
+            case HOLD_EFFECT_JABOCA_BERRY:  // consume and damage attacker if used physical move
+                if (IsBattlerAlive(battlerId)
+                  && TARGET_TURN_DAMAGED
+                  && !DoesSubstituteBlockMove(gBattlerAttacker, battlerId, gCurrentMove)
+                  && IS_MOVE_PHYSICAL(gCurrentMove))
```
Comment (ultima-soul) [inline src/battle_util.c:6152]:
```suggestion
                 && TARGET_TURN_DAMAGED
                 && !DoesSubstituteBlockMove(gBattlerAttacker, battlerId, gCurrentMove)
                 && IS_MOVE_PHYSICAL(gCurrentMove))
```

File: src/battle_util.c [hunk 4]
```diff
@@ -5989,6 +6145,50 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
...
+                    gBattlescriptCurrInstr = BattleScript_JabocaRowapBerryActivates;
+                    PREPARE_ITEM_BUFFER(gBattleTextBuff1, gLastUsedItem);
+                    RecordItemEffectBattle(battlerId, HOLD_EFFECT_ROCKY_HELMET);
+                }
+                break;
+            case HOLD_EFFECT_ROWAP_BERRY:  // consume and damage attacker if used special move
+                if (IsBattlerAlive(battlerId)
+                  && TARGET_TURN_DAMAGED
+                  && !DoesSubstituteBlockMove(gBattlerAttacker, battlerId, gCurrentMove)
+                  && IS_MOVE_SPECIAL(gCurrentMove))
```
Comment (ultima-soul) [inline src/battle_util.c:6171]:
```suggestion
                 && TARGET_TURN_DAMAGED
                 && !DoesSubstituteBlockMove(gBattlerAttacker, battlerId, gCurrentMove)
                 && IS_MOVE_SPECIAL(gCurrentMove))
```

**Review by ultima-soul — CHANGES_REQUESTED** [review]
- Magic Guard should prevent activation of Jaboca and Rowap Berry
- Change to item data for Sitrus Berry should happen in Item Expansion

Additionally something to be done by somebody:
- Hold effects for berries should be added to item data in Item Expansion

**Comment by ghoulslash** [comment]
>     * Magic Guard should prevent activation of Jaboca and Rowap Berry
Done
>     * Change to item data for Sitrus Berry should happen in Item Expansion
Removed that part. #1344 handles sitrus berry in IE too
> 
> Additionally something to be done by somebody:
> 
>     * Hold effects for berries should be added to item data in Item Expansion
Handled in #1415

File: src/battle_util.c [hunk 5]
```diff
@@ -5172,7 +5172,6 @@ enum
 bool32 HasEnoughHpToEatBerry(u32 battlerId, u32 hpFraction, u32 itemId)
 {
     bool32 isBerry = (ItemId_GetPocket(itemId) == POCKET_BERRIES);
```
Comment (ultima-soul) [inline src/battle_util.c:5174]:
```suggestion
    bool32 isBerry = (ItemId_GetPocket(itemId) == POCKET_BERRIES);

```

File: src/battle_util.c [hunk 6]
```diff
@@ -6212,9 +6211,10 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
                 break;
             case HOLD_EFFECT_JABOCA_BERRY:  // consume and damage attacker if used physical move
                 if (IsBattlerAlive(battlerId)
-                  && TARGET_TURN_DAMAGED
-                  && !DoesSubstituteBlockMove(gBattlerAttacker, battlerId, gCurrentMove)
-                  && IS_MOVE_PHYSICAL(gCurrentMove))
+                 && TARGET_TURN_DAMAGED
+                 && !DoesSubstituteBlockMove(gBattlerAttacker, battlerId, gCurrentMove)
+                 && IS_MOVE_PHYSICAL(gCurrentMove)
+                  && GetBattlerAbility(gBattlerAttacker) != ABILITY_MAGIC_GUARD)
```
Comment (ultima-soul) [inline src/battle_util.c:6217]:
```suggestion
                 && GetBattlerAbility(gBattlerAttacker) != ABILITY_MAGIC_GUARD)
```

File: src/battle_util.c [hunk 7]
```diff
@@ -6231,9 +6231,10 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
                 break;
             case HOLD_EFFECT_ROWAP_BERRY:  // consume and damage attacker if used special move
                 if (IsBattlerAlive(battlerId)
-                  && TARGET_TURN_DAMAGED
-                  && !DoesSubstituteBlockMove(gBattlerAttacker, battlerId, gCurrentMove)
-                  && IS_MOVE_SPECIAL(gCurrentMove))
+                 && TARGET_TURN_DAMAGED
+                 && !DoesSubstituteBlockMove(gBattlerAttacker, battlerId, gCurrentMove)
+                 && IS_MOVE_SPECIAL(gCurrentMove)
+                  && GetBattlerAbility(gBattlerAttacker) != ABILITY_MAGIC_GUARD)
```
Comment (ultima-soul) [inline src/battle_util.c:6237]:
```suggestion
                 && GetBattlerAbility(gBattlerAttacker) != ABILITY_MAGIC_GUARD)
```

File: src/battle_message.c [hunk 8]
```diff
@@ -691,13 +691,18 @@ static const u8 sText_BattlerAbilityRaisedStat[] = _("{B_SCR_ACTIVE_NAME_WITH_PR
 static const u8 sText_ASandstormKickedUp[] = _("A sandstorm kicked up!");
 static const u8 sText_PkmnsWillPerishIn3Turns[] = _("Both Pokémon will perish\nin three turns!");
 static const u8 sText_AbilityRaisedStatDrastically[] = _("{B_DEF_ABILITY} raised {B_DEF_NAME_WITH_PREFIX}'s\n{B_BUFF1} drastically!");
+static const u8 sText_AsOneEnters[] = _("{B_SCR_ACTIVE_NAME_WITH_PREFIX} has two Abilities!");
+static const u8 sText_CuriousMedicineEnters[] = _("{B_EFF_NAME_WITH_PREFIX}'s\nstat changes were reset!");
 static const u8 sText_CanActFaster[] = _("{B_ATK_NAME_WITH_PREFIX} can act faster,\nthanks to {B_LAST_ITEM}!");
 static const u8 sText_MicleBerryActivates[] = _("{B_SCR_ACTIVE_NAME_WITH_PREFIX} boosted the accuracy of its\nnext move using {B_LAST_ITEM}!");
 
+
```
Comment (ultima-soul) [inline src/battle_message.c:699]:
```suggestion
```

File: include/battle_scripts.h [hunk 9]
```diff
@@ -357,5 +359,11 @@ extern const u8 BattleScript_CottonDownActivates[];
 extern const u8 BattleScript_BallFetch[];
 extern const u8 BattleScript_SandSpitActivates[];
 extern const u8 BattleScript_PerishBodyActivates[];
+extern const u8 BattleScript_QuickClawActivation[];
+extern const u8 BattleScript_CustapBerryActivation[];
+extern const u8 BattleScript_MicleBerryActivateEnd2[];
+extern const u8 BattleScript_MicleBerryActivateRet[];
+extern const u8 BattleScript_JabocaRowapBerryActivates[];
+
```
[the remark this answers is not in the document]
Comment (ultima-soul) [reply include/battle_scripts.h:370]: Unresolved still.

**Review by ultima-soul — CHANGES_REQUESTED** [review]

**Comment by ghoulslash** [comment]
fixed formatting

**Review by ultima-soul — APPROVED** [review]

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**